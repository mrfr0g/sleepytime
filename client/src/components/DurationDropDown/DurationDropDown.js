import React from 'react';
import { DateTime } from 'luxon';
import { Select } from 'grommet';

/**
 * Returns a formatted string for the time object
 * @param {DateTime} time
 * @returns {String}
 */
function formatTimeSlot(time) {
  return `${time.toFormat('HH')}h:${time.toFormat('mm')}m`;
}

/**
 * Returns a list of time slots in minute increments
 * @param {Number} increments
 * @returns {Array} ["0h:30m", "1h:00m"]
 */
function getTimeSlots(increments = 30) {
  const slots = [];

  // Using Luxon here to calculate time from midnight, this may be a bit
  // more than we need, but it lets us easily deal with date math and
  // support more increments in the future
  let now = DateTime.fromISO('2021-01-01');

  // Generate slots until we reach the max hours per day
  while (now.toFormat('H') <= 23) {
    now = now.plus({
      minutes: increments,
    });

    // Break if we move past today
    if (now.toFormat('dd') !== '01') {
      break;
    }

    slots.push(formatTimeSlot(now));
  }

  return slots;
}

/**
 * Creates a dropdown with timestamps from 0-24 hours, in 30m increments (default)
 * @prop {Number} increments
 */
export function DurationDropDown({ increments = 30, ...props }) {
  const slots = getTimeSlots(increments);

  return <Select options={slots} {...props} />;
}
