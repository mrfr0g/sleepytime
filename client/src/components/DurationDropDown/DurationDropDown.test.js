import React from 'react';
import renderer from 'react-test-renderer';
import { DurationDropDown } from './DurationDropDown';

describe('DurationDropDown', () => {
  it('renders the same way every time', () => {
    const tree = renderer.create(<DurationDropDown increments={30} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
