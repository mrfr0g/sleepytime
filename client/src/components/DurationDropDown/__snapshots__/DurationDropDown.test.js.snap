// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationDropDown renders the same way every time 1`] = `
<button
  aria-label="Open Drop"
  className="StyledButton-sc-323bzc-0 fJlAlc Select__StyledSelectDropButton-sc-17idtfo-2 gbKWcB"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  type="button"
>
  <div
    className="StyledBox-sc-13pk1d4-0 fXagjJ"
  >
    <div
      className="StyledBox-sc-13pk1d4-0 fREhmO"
    >
      <div
        className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm"
      >
        <input
          autoComplete="off"
          className="StyledTextInput-sc-1x30a0s-0 defKuP Select__SelectTextInput-sc-17idtfo-0 lmyIUy"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={true}
          tabIndex="-1"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="StyledBox-sc-13pk1d4-0 dDjwWo"
      style={
        Object {
          "minWidth": "auto",
        }
      }
    >
      <svg
        aria-label="FormDown"
        className="StyledIcon-sc-ofa7kd-0 YhzyJ"
        viewBox="0 0 24 24"
      >
        <path
          d="m18 9-6 6-6-6"
          fill="none"
          stroke="#000"
          strokeWidth="2"
        />
      </svg>
    </div>
  </div>
</button>
`;
